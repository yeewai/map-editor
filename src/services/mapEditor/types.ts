export interface State {
    zoom: number,
    panV: number,
    panH: number
}
