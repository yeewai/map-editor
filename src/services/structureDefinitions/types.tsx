export interface StructureDefinition {
    id: string,
    name: string,
    description?: string,
    width: number,
    length: number,
    kind: string,
    imageUrl: string,
    imageWidth?: number,
    imageHeight?: number,
    image?: any
}

export const emptyStructureDefintion = {
    id: "",
    name: "",
    width: 1,
    length: 1,
    kind: "",
    imageUrl: ""
}

export interface State {
    items: StructureDefinition[],
    itemsRequested: number,
    isFetching: boolean,
    hasFetched: boolean,
    error?: string
}
