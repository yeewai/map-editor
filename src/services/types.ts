import { State as structureDefinitionType } from './structureDefinitions/types';
import { State as worldType } from './worlds/types';

import { defaultState as structureDefinitionState } from './structureDefinitions/reducer';
import { defaultState as worldState } from './worlds/reducer';

export interface StateTree {
    structureDefinitions: structureDefinitionType,
    worlds: worldType
}

export const defaultStateTree: StateTree = {
    structureDefinitions: structureDefinitionState,
    worlds: worldState
}

export interface Action {
    type: string,
    payload?: any
}

export const defaultAction:Action = {
    type: ""
}
