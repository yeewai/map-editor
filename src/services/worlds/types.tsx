import { structureDefinitionTypes } from 'services/structureDefinitions';

export interface Structure {
    name: string,
    description?: string,
    definitionId: string,
    definition?: structureDefinitionTypes.StructureDefinition
    xposition: number,
    yposition: number
}

export const emptyStructure = {
    name: "",
    definitionId: structureDefinitionTypes.emptyStructureDefintion.id,
    definition: structureDefinitionTypes.emptyStructureDefintion,
    xposition: 1,
    yposition: 1
}

export interface World {
    id: string,
    name: string,
    key: string,
    description?: string,
    width: number,
    length: number,
    createdAt: number,
    saveMessage?: string,
    isPublished: boolean,
    structures?: Structure[],
    nullStructureId: string
}

export const emptyWorld = {
    id: "",
    name: "",
    key: "",
    width: 1,
    length: 1,
    createdAt: 1,
    isPublished: false,
    structures: [ emptyStructure ],
    nullStructureId: structureDefinitionTypes.emptyStructureDefintion.id
}

export interface WorldWithBoard extends World {
    board: Structure[][],
    uniqueStructures: structureDefinitionTypes.StructureDefinition[]
}

export const emptyWorldWithBoard = {
    ...emptyWorld,
    board: [
        [ emptyStructure ]
    ],
    uniqueStructures: [ structureDefinitionTypes.emptyStructureDefintion ]
}

export interface State {
    items: World[],
    isFetching: boolean,
    hasFetched: boolean,
    error?: string
}
