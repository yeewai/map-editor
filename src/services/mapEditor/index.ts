import reducer from "./reducer";
import * as mapEditorActions from "./actions";
import * as mapEditorSelectors from "./selectors";
import * as mapEditorTypes from "./types";

export { mapEditorActions, mapEditorSelectors, mapEditorTypes };
export default reducer;
