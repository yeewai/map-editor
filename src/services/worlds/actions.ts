import { ActionCreator, Dispatch } from 'redux';
import { ThunkAction } from 'redux-thunk';
import { World } from './types';
import { StateTree, Action } from 'services/types';
import * as api from 'api';
import _ from 'lodash';

import { modalActions } from '@evercourse/ever-modal';
import { structureDefinitionsSelectors } from 'services/structureDefinitions';

export const requestWorlds: ActionCreator<Action> = () => ({
    type: 'world/REQUEST'
});

export const receiveWorlds: ActionCreator<Action> = ( worlds: World[] )=> ({
    type: 'world/RECEIVE',
    payload: worlds
});

export const setError: ActionCreator<Action> = (jqXHR: any, textStatus: string, errorThrown: any) => ({
    type: 'world/SET_ERROR',
    payload: { jqXHR, textStatus, errorThrown }
});

//I have no idea what the last return type is so I set it to any.........
export const fetchWorlds: ActionCreator<ThunkAction<any, StateTree, void>> = () => {
    return (dispatch: Dispatch<StateTree>): any => {
        dispatch(requestWorlds());
        return api.getWorlds().then (
            resp => dispatch(receiveWorlds(resp)),
            (jqXHR, textStatus, errorThrown) =>{
                 dispatch(setError(jqXHR, textStatus, errorThrown))
             }
        );
    };
};

const putWorld: ActionCreator<Action> = () => ({
    type: 'world/PUT'
});

export const updateWorld: ActionCreator<ThunkAction<any, StateTree, void>> = (values) => {
    return (dispatch: Dispatch<StateTree>, getState: () => StateTree ) => {

        dispatch(putWorld());
        return dispatch(processWorld(values));
    };
}

const createWorld: ActionCreator<Action> = () => ({
    type: 'world/CREATE'
});

export const addWorld: ActionCreator<ThunkAction<any, StateTree, void>> = (values) => {
    return (dispatch: Dispatch<StateTree>, getState: () => StateTree ) => {

        dispatch(createWorld());

        return dispatch(processWorld(values));
    };
}

// const processingWorld: ActionCreator<Action> = () => ({
//     type: 'world/PROCESS'
// });
export const processWorld: ActionCreator<ThunkAction<any, StateTree, void>> = (values) => {
    return (dispatch: Dispatch<StateTree>, getState: () => StateTree ) => {

        return api.createWorld(values).then(
            resp => {
                dispatch(modalActions.hideModal());
                dispatch(fetchWorlds());
            },
            (jqXHR, textStatus, errorThrown) => dispatch(setError(jqXHR, textStatus, errorThrown))
        );
    };
}
