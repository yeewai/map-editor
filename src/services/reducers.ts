import { routerReducer } from 'react-router-redux';
import { combineReducers } from 'redux';
import { modalReducer } from 'ever-modal';
import { reducer as formReducer } from 'redux-form'

import { StateTree } from './types';
import structureDefinitions from './structureDefinitions';
import worlds from './worlds';

const rootReducer = combineReducers<StateTree>({
    router: routerReducer,
    modals: modalReducer,
    form: formReducer,
    structureDefinitions,
    worlds

});

export default rootReducer;
