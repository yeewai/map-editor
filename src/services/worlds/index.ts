import reducer from "./reducer";
import * as worldActions from "./actions";
import * as worldSelectors from "./selectors";
import * as worldTypes from "./types";

export { worldActions, worldSelectors, worldTypes };
export default reducer;
