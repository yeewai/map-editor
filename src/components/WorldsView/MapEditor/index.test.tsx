import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import sinon from 'sinon';

import { getMockRouterProps } from 'helpers/test/react-router';
import { MapEditor, mapStateToProps } from './index';
import { worldSelectors, worldTypes } from 'services/worlds';
import { defaultStateTree } from 'services/types';

Enzyme.configure({ adapter: new Adapter() });

describe("Map Editor", () => {
    let sandbox, wrapper, routerProps, world;

    beforeEach(() => {
        sandbox = sinon.createSandbox();
        routerProps = getMockRouterProps<any>(null);
        world = worldTypes.emptyWorldWithBoard;
    });

    afterEach(() => {
        sandbox.restore();
    });

    it ("renders", () => {
        wrapper = shallow(<MapEditor world={world} />);
        expect(wrapper).toMatchSnapshot();
    });

    it ("renders an error if there is no world", () => {
        wrapper = shallow(<MapEditor world={null} />);
        expect(wrapper.find("Alert").exists()).toBe(true);
    })

    it ("maps state to props", () => {
        routerProps.match.params = { id: "a"};
        sandbox.stub(worldSelectors, 'getWorldWithBoardByKey').returns(world)

        expect(mapStateToProps(defaultStateTree, routerProps )).toEqual({world});
    })

});
