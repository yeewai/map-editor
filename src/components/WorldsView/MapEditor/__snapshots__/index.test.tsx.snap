// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map Editor renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    world={
        Object {
            "board": Array [
              Array [
                Object {
                  "definition": Object {
                    "id": "",
                    "imageUrl": "",
                    "kind": "",
                    "length": 1,
                    "name": "",
                    "width": 1,
                  },
                  "definitionId": "",
                  "name": "",
                  "xposition": 1,
                  "yposition": 1,
                },
              ],
            ],
            "createdAt": 1,
            "id": "",
            "isPublished": false,
            "key": "",
            "length": 1,
            "name": "",
            "nullStructureId": "",
            "structures": Array [
              Object {
                "definition": Object {
                  "id": "",
                  "imageUrl": "",
                  "kind": "",
                  "length": 1,
                  "name": "",
                  "width": 1,
                },
                "definitionId": "",
                "name": "",
                "xposition": 1,
                "yposition": 1,
              },
            ],
            "uniqueStructures": Array [
              Object {
                "id": "",
                "imageUrl": "",
                "kind": "",
                "length": 1,
                "name": "",
                "width": 1,
              },
            ],
            "width": 1,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2 />,
        <Connect(Component) />,
        <Connect(Component)
          world={
                    Object {
                              "board": Array [
                                Array [
                                  Object {
                                    "definition": Object {
                                      "id": "",
                                      "imageUrl": "",
                                      "kind": "",
                                      "length": 1,
                                      "name": "",
                                      "width": 1,
                                    },
                                    "definitionId": "",
                                    "name": "",
                                    "xposition": 1,
                                    "yposition": 1,
                                  },
                                ],
                              ],
                              "createdAt": 1,
                              "id": "",
                              "isPublished": false,
                              "key": "",
                              "length": 1,
                              "name": "",
                              "nullStructureId": "",
                              "structures": Array [
                                Object {
                                  "definition": Object {
                                    "id": "",
                                    "imageUrl": "",
                                    "kind": "",
                                    "length": 1,
                                    "name": "",
                                    "width": 1,
                                  },
                                  "definitionId": "",
                                  "name": "",
                                  "xposition": 1,
                                  "yposition": 1,
                                },
                              ],
                              "uniqueStructures": Array [
                                Object {
                                  "id": "",
                                  "imageUrl": "",
                                  "kind": "",
                                  "length": 1,
                                  "name": "",
                                  "width": 1,
                                },
                              ],
                              "width": 1,
                            }
          }
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
        },
        "ref": null,
        "rendered": "",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "world": Object {
            "board": Array [
              Array [
                Object {
                  "definition": Object {
                    "id": "",
                    "imageUrl": "",
                    "kind": "",
                    "length": 1,
                    "name": "",
                    "width": 1,
                  },
                  "definitionId": "",
                  "name": "",
                  "xposition": 1,
                  "yposition": 1,
                },
              ],
            ],
            "createdAt": 1,
            "id": "",
            "isPublished": false,
            "key": "",
            "length": 1,
            "name": "",
            "nullStructureId": "",
            "structures": Array [
              Object {
                "definition": Object {
                  "id": "",
                  "imageUrl": "",
                  "kind": "",
                  "length": 1,
                  "name": "",
                  "width": 1,
                },
                "definitionId": "",
                "name": "",
                "xposition": 1,
                "yposition": 1,
              },
            ],
            "uniqueStructures": Array [
              Object {
                "id": "",
                "imageUrl": "",
                "kind": "",
                "length": 1,
                "name": "",
                "width": 1,
              },
            ],
            "width": 1,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "article",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2 />,
          <Connect(Component) />,
          <Connect(Component)
            world={
                        Object {
                                    "board": Array [
                                      Array [
                                        Object {
                                          "definition": Object {
                                            "id": "",
                                            "imageUrl": "",
                                            "kind": "",
                                            "length": 1,
                                            "name": "",
                                            "width": 1,
                                          },
                                          "definitionId": "",
                                          "name": "",
                                          "xposition": 1,
                                          "yposition": 1,
                                        },
                                      ],
                                    ],
                                    "createdAt": 1,
                                    "id": "",
                                    "isPublished": false,
                                    "key": "",
                                    "length": 1,
                                    "name": "",
                                    "nullStructureId": "",
                                    "structures": Array [
                                      Object {
                                        "definition": Object {
                                          "id": "",
                                          "imageUrl": "",
                                          "kind": "",
                                          "length": 1,
                                          "name": "",
                                          "width": 1,
                                        },
                                        "definitionId": "",
                                        "name": "",
                                        "xposition": 1,
                                        "yposition": 1,
                                      },
                                    ],
                                    "uniqueStructures": Array [
                                      Object {
                                        "id": "",
                                        "imageUrl": "",
                                        "kind": "",
                                        "length": 1,
                                        "name": "",
                                        "width": 1,
                                      },
                                    ],
                                    "width": 1,
                                  }
            }
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "world": Object {
              "board": Array [
                Array [
                  Object {
                    "definition": Object {
                      "id": "",
                      "imageUrl": "",
                      "kind": "",
                      "length": 1,
                      "name": "",
                      "width": 1,
                    },
                    "definitionId": "",
                    "name": "",
                    "xposition": 1,
                    "yposition": 1,
                  },
                ],
              ],
              "createdAt": 1,
              "id": "",
              "isPublished": false,
              "key": "",
              "length": 1,
              "name": "",
              "nullStructureId": "",
              "structures": Array [
                Object {
                  "definition": Object {
                    "id": "",
                    "imageUrl": "",
                    "kind": "",
                    "length": 1,
                    "name": "",
                    "width": 1,
                  },
                  "definitionId": "",
                  "name": "",
                  "xposition": 1,
                  "yposition": 1,
                },
              ],
              "uniqueStructures": Array [
                Object {
                  "id": "",
                  "imageUrl": "",
                  "kind": "",
                  "length": 1,
                  "name": "",
                  "width": 1,
                },
              ],
              "width": 1,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "article",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
